Teuchos: strange warning under CUDA

Here is the warning:
```
246/442] Building CXX object packages/tpetra/core/src.../tpetra.dir/Tpetra_CrsMatrix_DOUBLE_INT_INT_CUDA.cpp.o
/home/jhu/software/trilinos/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp: In instantiation of ‘Teuchos::SerialDenseMatrix<OrdinalType, ScalarType>& Teuchos::SerialDenseMatrix<OrdinalType, ScalarType>::operator=(const Teuchos::SerialDenseMatrix<OrdinalType, ScalarType>&) [with OrdinalType = int; ScalarType = double]’:
/home/jhu/software/trilinos/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseSolver.hpp:690:11:   required from ‘int Teuchos::SerialDenseSolver<OrdinalType, ScalarType>::solve() [with OrdinalType = int; ScalarType = double]’
/home/jhu/software/trilinos/Trilinos/packages/muelu/src/Transfers/BlackBox/MueLu_BlackBoxPFactory_def.hpp:1637:1:   required from ‘void MueLu::BlackBoxPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::ComputeLocalEntries(const Teuchos::RCP<const Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >&, Teuchos::Array<T>, Teuchos::Array<T>, LocalOrdinal, Teuchos::Array<T>, Teuchos::Array<T>, LocalOrdinal, Teuchos::Array<T>, Teuchos::Array<GlobalOrdinal>, Teuchos::Array<GlobalOrdinal>, Teuchos::Array<T>, Teuchos::Array<bool>, Teuchos::Array<int>, std::__cxx11::string, std::__cxx11::string, Teuchos::Array<T>, LocalOrdinal, Teuchos::Array<GlobalOrdinal>, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::SerialDenseMatrix<LocalOrdinal, Scalar>&, Teuchos::Array<T>&, Teuchos::Array<T>&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = int; Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/jhu/software/trilinos/Trilinos/packages/muelu/src/Utils/MueLu_ETI_4arg.hpp:42:23:   required from here
/home/jhu/software/trilinos/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp:659:14: warning: non-constant array new length must be specified without parentheses around the type-id [-Wvla]
         values_ = new ScalarType[newsize];
             ~^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

but as you can see, there are no parentheses.

@trilinos/teuchos

@jhux2, thanks for pointing that out.  I see them in the `Trilinos-atdm-waterman-cuda-9.2-debug` build on CDash [here](https://testing.sandia.gov/cdash-dev-view/viewBuildError.php?type=1&buildid=4063676) showing:

```
/home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-debug/SRC_AND_BUILD/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp: In instantiation of ‘Teuchos::SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const Teuchos::SerialDenseMatrix<OrdinalType, ScalarType>&, Teuchos::ETransp) [with OrdinalType = int; ScalarType = double]’:
/home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-debug/SRC_AND_BUILD/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialQRDenseSolver.hpp:836:18:   required from ‘int Teuchos::SerialQRDenseSolver<OrdinalType, ScalarType>::formQ() [with OrdinalType = int; ScalarType = double]’
/home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-debug/SRC_AND_BUILD/Trilinos/packages/teuchos/numerics/test/DenseMatrix/cxx_main_qr_solver.cpp:171:28:   required from here
[CTest: warning matched] /home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-debug/SRC_AND_BUILD/Trilinos/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp:455:14: warning: non-constant array new length must be specified without parentheses around the type-id [-Wvla]
         values_ = new ScalarType[newsize];
```

@hkthorn, mind if we try replacing raw new/delete with something like `Teuchos::Array<T>`?

--- a/packages/belos/epetra/src/BelosEpetraAdapter.cpp
+++ b/packages/belos/epetra/src/BelosEpetraAdapter.cpp
@@ -88,11 +88,11 @@ namespace Belos {
       static RCP<std::ostream> errStream;
 
       if (errStream.is_null()) {
-	if (Op.Comm().MyPID() == 0) {
-	  errStream = rcpFromRef (std::cerr);
-	} else {
-	  errStream = rcp (new Teuchos::oblackholestream);
-	}
+        if (Op.Comm().MyPID() == 0) {
+          errStream = rcpFromRef (std::cerr);
+        } else {
+          errStream = rcp (new Teuchos::oblackholestream);
+        }
       }
       return errStream;
     }
@@ -103,14 +103,14 @@ namespace Belos {
     etransToString (const ETrans trans)
     {
       if (trans == NOTRANS) {
-	return "NOTRANS";
+        return "NOTRANS";
       } else if (trans == TRANS) {
-	return "TRANS";
+        return "TRANS";
       } else if (trans == CONJTRANS) {
-	return "CONJTRANS";
+        return "CONJTRANS";
       }
       TEUCHOS_TEST_FOR_EXCEPTION(true, std::invalid_argument,
-			 "Invalid ETrans value trans = " << trans << ".");
+                         "Invalid ETrans value trans = " << trans << ".");
     }
 
     /// \fn implementsApplyTranspose
@@ -118,7 +118,7 @@ namespace Belos {
     ///
     /// Epetra_Operator instances are not required to implement
     /// applying the transpose (or conjugate transpose, if
-    /// applicable).  This function lets you ask.  
+    /// applicable).  This function lets you ask.
     ///
     /// We need this function because Epetra_Operator doesn't have a
     /// direct query method.  We query by asking it to set using the
@@ -133,7 +133,7 @@ namespace Belos {
     /// We use this function multiple times in this file, so it's
     /// worthwhile breaking it out into a separate function, rather
     /// than copying and pasting.
-    bool 
+    bool
     implementsApplyTranspose (const Epetra_Operator& Op)
     {
       // We must first check Op's current transpose state, so that we
@@ -143,15 +143,15 @@ namespace Belos {
       // If Op is already transposed, then obviously it implements
       // applying the transpose.
       if (transposed) {
-	return true;
+        return true;
       }
 
       // SetUseTranspose() is a nonconst method, so we have to cast
       // away const-ness of Op in order to call it.  Epetra_Operator
       // gives us no other way to query whether it knows how to apply
       // the transpose.
-      const int transposeFailed = 
-	const_cast<Epetra_Operator&>(Op).SetUseTranspose (true);
+      const int transposeFailed =
+        const_cast<Epetra_Operator&>(Op).SetUseTranspose (true);
 
       // SetUseTranspose() follows the POSIX convention that returning
       // zero means success.  If SetUseTranspose() failed, the best we can
@@ -159,35 +159,35 @@ namespace Belos {
       // reasonable implementation of SetUseTranspose() should not change
       // the state of the operator if setting the transpose failed.
       if (transposeFailed != 0) {
-	// Make sure that the failed call didn't managed to change the
-	// state of Op.  It shouldn't have, but if it did, then Op is
-	// broken.
-	const bool currentTransposedState = Op.UseTranspose();
-	TEUCHOS_TEST_FOR_EXCEPTION(currentTransposedState != transposed, std::logic_error,
-			   "implementsApplyTranspose: Op.SetUseTranspose(true) "
-			   "failed (with error code " << transposeFailed << "), "
-			   "but nevertheless managed to change the transposed "
-			   "state of Op.  On input, Op.UseTranspose() = " 
-			   << transposed << ", but now, Op.UseTranspose() = " 
-			   << currentTransposedState << ".  This likely indicates"
-			   " a bug in the implementation of the operator Op.");
-	return false;
+        // Make sure that the failed call didn't managed to change the
+        // state of Op.  It shouldn't have, but if it did, then Op is
+        // broken.
+        const bool currentTransposedState = Op.UseTranspose();
+        TEUCHOS_TEST_FOR_EXCEPTION(currentTransposedState != transposed, std::logic_error,
+                           "implementsApplyTranspose: Op.SetUseTranspose(true) "
+                           "failed (with error code " << transposeFailed << "), "
+                           "but nevertheless managed to change the transposed "
+                           "state of Op.  On input, Op.UseTranspose() = "
+                           << transposed << ", but now, Op.UseTranspose() = "
+                           << currentTransposedState << ".  This likely indicates"
+                           " a bug in the implementation of the operator Op.");
+        return false;
       }
 
       // Now we need to restore the original transpose state.  We can
       // do this unconditionally because if we're here, Op's state is
-      // not transposed.  
+      // not transposed.
       {
-	const int errcode = 
-	  const_cast<Epetra_Operator&>(Op).SetUseTranspose (false);
-	TEUCHOS_TEST_FOR_EXCEPTION(errcode != 0, std::logic_error, 
-			   "implementsApplyTranspose: Op.SetUseTranspose(true) "
-			   "succeeded, but Op.SetUseTranspose(false) failed with "
-			   "error code " << errcode << ".  This likely indicates "
-			   "a bug in the implementation of the operator Op.");
+        const int errcode =
+          const_cast<Epetra_Operator&>(Op).SetUseTranspose (false);
+        TEUCHOS_TEST_FOR_EXCEPTION(errcode != 0, std::logic_error,
+                           "implementsApplyTranspose: Op.SetUseTranspose(true) "
+                           "succeeded, but Op.SetUseTranspose(false) failed with "
+                           "error code " << errcode << ".  This likely indicates "
+                           "a bug in the implementation of the operator Op.");
       }
       // If we made it this far, then Op implements transpose.
-      return true; 
+      return true;
     }
 
     /// \class EpetraOperatorTransposeScopeGuard
@@ -201,7 +201,7 @@ namespace Belos {
     private:
       //! Reference to the Epetra_Operator instance to protect.
       const Epetra_Operator& Op_;
-      
+
       //! The original transpose state of the operator.
       const bool originalTransposeFlag_;
 
@@ -219,134 +219,134 @@ namespace Belos {
       ///
       /// \param trans [in] True if we want to set Op to apply the
       ///   transpose, else false.
-      EpetraOperatorTransposeScopeGuard (const Epetra_Operator& Op, 
-					 const ETrans trans)
-	: Op_ (Op),
-	  originalTransposeFlag_ (Op.UseTranspose ()),
-	  newTransposeFlag_ (trans != NOTRANS)
+      EpetraOperatorTransposeScopeGuard (const Epetra_Operator& Op,
+                                         const ETrans trans)
+        : Op_ (Op),
+          originalTransposeFlag_ (Op.UseTranspose ()),
+          newTransposeFlag_ (trans != NOTRANS)
       {
-	// If necessary, set (or unset) the transpose flag to the value
-	// corresponding to 'trans'.
-	if (newTransposeFlag_ != originalTransposeFlag_) {
+        // If necessary, set (or unset) the transpose flag to the value
+        // corresponding to 'trans'.
+        if (newTransposeFlag_ != originalTransposeFlag_) {
 
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
-	  using std::endl;
-	  std::ostream& err = *(getErrStream (Op_));
-	  err << "---- EpetraOperatorTransposeScopeGuard constructor:" << endl
-	      << "------ Toggling transpose flag from " << originalTransposeFlag_ 
-	      << " to " << newTransposeFlag_ << endl;
+          using std::endl;
+          std::ostream& err = *(getErrStream (Op_));
+          err << "---- EpetraOperatorTransposeScopeGuard constructor:" << endl
+              << "------ Toggling transpose flag from " << originalTransposeFlag_
+              << " to " << newTransposeFlag_ << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
-	  // Toggle the transpose flag.  The destructor will restore
-	  // its original value.
-	  const int info = 
-	    const_cast<Epetra_Operator &>(Op_).SetUseTranspose (newTransposeFlag_);
-	  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure,
-			     "Toggling the transpose flag of the operator failed, "
-			     "returning a nonzero error code of " << info 
-			     << ".  That probably means that the Epetra_Operator "
-			     "subclass instance doesn't know how to apply its "
-			     "transpose.  Are you perhaps using a subclass of "
-			     "Epetra_Operator for which applying the transpose "
-			     "is not implemented?  Anyway, just to help with "
-			     "debugging, you specified trans=" 
-			     << etransToString (trans)
-			     << ", and the original operator was set "
-			     << (originalTransposeFlag_ ? "" : "NOT ")
-			     << "to apply the transpose.");
-	}
+          // Toggle the transpose flag.  The destructor will restore
+          // its original value.
+          const int info =
+            const_cast<Epetra_Operator &>(Op_).SetUseTranspose (newTransposeFlag_);
+          TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure,
+                             "Toggling the transpose flag of the operator failed, "
+                             "returning a nonzero error code of " << info
+                             << ".  That probably means that the Epetra_Operator "
+                             "subclass instance doesn't know how to apply its "
+                             "transpose.  Are you perhaps using a subclass of "
+                             "Epetra_Operator for which applying the transpose "
+                             "is not implemented?  Anyway, just to help with "
+                             "debugging, you specified trans="
+                             << etransToString (trans)
+                             << ", and the original operator was set "
+                             << (originalTransposeFlag_ ? "" : "NOT ")
+                             << "to apply the transpose.");
+        }
       }
 
       ~EpetraOperatorTransposeScopeGuard ()
       {
-	// SetUseTranspose() changes the state of the operator, so if
-	// applicable, we have to change the state back.
-	if (newTransposeFlag_ != originalTransposeFlag_) {
+        // SetUseTranspose() changes the state of the operator, so if
+        // applicable, we have to change the state back.
+        if (newTransposeFlag_ != originalTransposeFlag_) {
 
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
-	  using std::endl;
-	  std::ostream& err = *(getErrStream (Op_));
-	  err << "---- EpetraOperatorTransposeScopeGuard destructor:" << endl
-	      << "------ Toggling transpose flag from " << newTransposeFlag_
-	      << " to " << originalTransposeFlag_ << endl;
+          using std::endl;
+          std::ostream& err = *(getErrStream (Op_));
+          err << "---- EpetraOperatorTransposeScopeGuard destructor:" << endl
+              << "------ Toggling transpose flag from " << newTransposeFlag_
+              << " to " << originalTransposeFlag_ << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
-	  const int info = 
-	    const_cast<Epetra_Operator &>(Op_).SetUseTranspose (originalTransposeFlag_);
-	  TEUCHOS_TEST_FOR_TERMINATION(info != 0,
-			     "Resetting the original "
-			     "transpose flag value of the Epetra_Operator failed, "
-			     "returning a nonzero error code of " << info << ".  "
-			     "This is an unusual error, since we were able to call "
-			     "its SetUseTranspose() method successfully before.  "
-			     "This suggests a bug in the subclass of Epetra_Operator "
-			     "which you are currently using.  This is probably not a "
-			     "Belos bug.");
-	}
+          const int info =
+            const_cast<Epetra_Operator &>(Op_).SetUseTranspose (originalTransposeFlag_);
+          TEUCHOS_TEST_FOR_TERMINATION(info != 0,
+                             "Resetting the original "
+                             "transpose flag value of the Epetra_Operator failed, "
+                             "returning a nonzero error code of " << info << ".  "
+                             "This is an unusual error, since we were able to call "
+                             "its SetUseTranspose() method successfully before.  "
+                             "This suggests a bug in the subclass of Epetra_Operator "
+                             "which you are currently using.  This is probably not a "
+                             "Belos bug.");
+        }
 
-	// Make sure that the transpose flag has its original value.  If
-	// not, we throw std::logic_error instead of EpetraOpFailure, since
-	// that's definitely a bug.  It's safe to do this check whether or
-	// not we actually had to toggle the transpose flag.  Any reasonable
-	// implementation of Epetra_Operator should make calling
-	// UseTranspose() cheap.
-	//
-	// Note to code maintainers: The reason we capture the value of
-	// UseTranspose() instead of just calling it twice, is that if the
-	// exception test really does trigger, then there is something
-	// fundamentally wrong.  If something is that wrong, then we want to
-	// avoid further calls to the operator's methods.  For example, it
-	// could be that the UseTranspose() method is erroneously changing
-	// the value of the flag, so if we call that method twice, it might
-	// have the right value on the second call.  This would make the
-	// resulting exception message confusing.
-	const bool finalTransposeFlag = Op_.UseTranspose ();
-	TEUCHOS_TEST_FOR_TERMINATION(originalTransposeFlag_ != finalTransposeFlag,
-			   "Belos::OperatorTraits::Apply: The value of the "
-			   "Epetra_Operator's transpose flag changed unexpectedly!"
-			   "  The original value at the top of this method was "
-			   << originalTransposeFlag_ << ", and its new value is "
-			   << finalTransposeFlag << ".  This suggests either a "
-			   "bug in Belos (in the implementation of this routine), "
-			   "or a bug in the operator.");
+        // Make sure that the transpose flag has its original value.  If
+        // not, we throw std::logic_error instead of EpetraOpFailure, since
+        // that's definitely a bug.  It's safe to do this check whether or
+        // not we actually had to toggle the transpose flag.  Any reasonable
+        // implementation of Epetra_Operator should make calling
+        // UseTranspose() cheap.
+        //
+        // Note to code maintainers: The reason we capture the value of
+        // UseTranspose() instead of just calling it twice, is that if the
+        // exception test really does trigger, then there is something
+        // fundamentally wrong.  If something is that wrong, then we want to
+        // avoid further calls to the operator's methods.  For example, it
+        // could be that the UseTranspose() method is erroneously changing
+        // the value of the flag, so if we call that method twice, it might
+        // have the right value on the second call.  This would make the
+        // resulting exception message confusing.
+        const bool finalTransposeFlag = Op_.UseTranspose ();
+        TEUCHOS_TEST_FOR_TERMINATION(originalTransposeFlag_ != finalTransposeFlag,
+                           "Belos::OperatorTraits::Apply: The value of the "
+                           "Epetra_Operator's transpose flag changed unexpectedly!"
+                           "  The original value at the top of this method was "
+                           << originalTransposeFlag_ << ", and its new value is "
+                           << finalTransposeFlag << ".  This suggests either a "
+                           "bug in Belos (in the implementation of this routine), "
+                           "or a bug in the operator.");
       }
     };
 
   } // namespace (anonymous)
 
-  
-EpetraMultiVec::EpetraMultiVec (const Epetra_BlockMap& Map_in, 
-				double * array, 
-			        const int numvecs, 
-				const int stride)
-  : Epetra_MultiVector(BELOSEPETRACOPY, Map_in, array, stride, numvecs) 
+
+EpetraMultiVec::EpetraMultiVec (const Epetra_BlockMap& Map_in,
+                                double * array,
+                                const int numvecs,
+                                const int stride)
+  : Epetra_MultiVector(BELOSEPETRACOPY, Map_in, array, stride, numvecs)
 {
 }
 
 
-EpetraMultiVec::EpetraMultiVec (const Epetra_BlockMap& Map_in, 
-				const int numvecs, 
-				bool zeroOut)
-  : Epetra_MultiVector(Map_in, numvecs, zeroOut) 
+EpetraMultiVec::EpetraMultiVec (const Epetra_BlockMap& Map_in,
+                                const int numvecs,
+                                bool zeroOut)
+  : Epetra_MultiVector(Map_in, numvecs, zeroOut)
 {
 }
 
 
-EpetraMultiVec::EpetraMultiVec (Epetra_DataAccess CV_in, 
-				const Epetra_MultiVector& P_vec,
-				const std::vector<int>& index )
+EpetraMultiVec::EpetraMultiVec (Epetra_DataAccess CV_in,
+                                const Epetra_MultiVector& P_vec,
+                                const std::vector<int>& index )
   : Epetra_MultiVector(CV_in, P_vec, &(const_cast<std::vector<int> &>(index))[0], index.size())
 {
 }
 
 
 EpetraMultiVec::EpetraMultiVec(const Epetra_MultiVector& P_vec)
-  : Epetra_MultiVector(P_vec) 
+  : Epetra_MultiVector(P_vec)
 {
 }
 
 
-EpetraMultiVec::~EpetraMultiVec() 
+EpetraMultiVec::~EpetraMultiVec()
 {
 }
 //
@@ -383,45 +383,45 @@ MultiVec<double>* EpetraMultiVec::CloneCopy ( const std::vector<int>& index ) co
 }
 
 
-MultiVec<double>* EpetraMultiVec::CloneViewNonConst ( const std::vector<int>& index ) 
+MultiVec<double>* EpetraMultiVec::CloneViewNonConst ( const std::vector<int>& index )
 {
   EpetraMultiVec * ptr_apv = new EpetraMultiVec(BELOSEPETRAVIEW, *this, index);
   return ptr_apv; // safe upcast.
 }
-  
 
-const MultiVec<double>* 
+
+const MultiVec<double>*
 EpetraMultiVec::CloneView (const std::vector<int>& index) const
 {
   EpetraMultiVec * ptr_apv = new EpetraMultiVec(BELOSEPETRAVIEW, *this, index);
   return ptr_apv; // safe upcast.
 }
-  
 
-void 
-EpetraMultiVec::SetBlock (const MultiVec<double>& A, 
-			  const std::vector<int>& index) 
-{	
+
+void
+EpetraMultiVec::SetBlock (const MultiVec<double>& A,
+                          const std::vector<int>& index)
+{
   EpetraMultiVec temp_vec(BELOSEPETRAVIEW, *this, index);
-  
+
   int numvecs = index.size();
   if ( A.GetNumberVecs() != numvecs ) {
     std::vector<int> index2( numvecs );
     for(int i=0; i<numvecs; i++)
       index2[i] = i;
-    EpetraMultiVec *tmp_vec = 
-      dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A)); 
+    EpetraMultiVec *tmp_vec =
+      dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A));
     TEUCHOS_TEST_FOR_EXCEPTION(tmp_vec==NULL, EpetraMultiVecFailure,
-		       "Belos::EpetraMultiVec::SetBlock: Dynamic cast from "
-		       "Belos::MultiVec<double> to Belos::EpetraMultiVec "
-		       "failed.");
+                       "Belos::EpetraMultiVec::SetBlock: Dynamic cast from "
+                       "Belos::MultiVec<double> to Belos::EpetraMultiVec "
+                       "failed.");
     EpetraMultiVec A_vec(BELOSEPETRAVIEW, *tmp_vec, index2);
     temp_vec.MvAddMv( 1.0, A_vec, 0.0, A_vec );
   }
   else {
     temp_vec.MvAddMv( 1.0, A, 0.0, A );
   }
-}								
+}
 
 //-------------------------------------------------------------
 //
@@ -429,19 +429,19 @@ EpetraMultiVec::SetBlock (const MultiVec<double>& A,
 //
 //-------------------------------------------------------------
 
-void EpetraMultiVec::MvTimesMatAddMv ( const double alpha, const MultiVec<double>& A, 
-				       const Teuchos::SerialDenseMatrix<int,double>& B, const double beta ) 
+void EpetraMultiVec::MvTimesMatAddMv ( const double alpha, const MultiVec<double>& A,
+                                       const Teuchos::SerialDenseMatrix<int,double>& B, const double beta )
 {
   Epetra_LocalMap LocalMap(B.numRows(), 0, Map().Comm());
   Epetra_MultiVector B_Pvec(BELOSEPETRAVIEW, LocalMap, B.values(), B.stride(), B.numCols());
-  
-  EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A)); 
+
+  EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A));
   TEUCHOS_TEST_FOR_EXCEPTION(A_vec==NULL, EpetraMultiVecFailure,
                      "Belos::EpetraMultiVec::MvTimesMatAddMv cast from Belos::MultiVec<> to Belos::EpetraMultiVec failed.");
-  
+
   int info = Multiply( 'N', 'N', alpha, *A_vec, B_Pvec, beta );
-  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure, 
-		     "Belos::EpetraMultiVec::MvTimesMatAddMv call to Multiply() returned a nonzero value.");
+  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure,
+                     "Belos::EpetraMultiVec::MvTimesMatAddMv call to Multiply() returned a nonzero value.");
 
 }
 
@@ -450,9 +450,9 @@ void EpetraMultiVec::MvTimesMatAddMv ( const double alpha, const MultiVec<double
 // *this <- alpha * A + beta * B
 //
 //-------------------------------------------------------------
-  
-void EpetraMultiVec::MvAddMv ( const double alpha , const MultiVec<double>& A, 
-			       const double beta, const MultiVec<double>& B) 
+
+void EpetraMultiVec::MvAddMv ( const double alpha , const MultiVec<double>& A,
+                               const double beta, const MultiVec<double>& B)
 {
   EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A));
   TEUCHOS_TEST_FOR_EXCEPTION( A_vec==NULL, EpetraMultiVecFailure,
@@ -460,10 +460,10 @@ void EpetraMultiVec::MvAddMv ( const double alpha , const MultiVec<double>& A,
   EpetraMultiVec *B_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(B));
   TEUCHOS_TEST_FOR_EXCEPTION( B_vec==NULL, EpetraMultiVecFailure,
                      "Belos::EpetraMultiVec::MvAddMv cast from Belos::MultiVec<> to Belos::EpetraMultiVec failed.");
-  
+
   int info = Update( alpha, *A_vec, beta, *B_vec, 0.0 );
-  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure, 
-		     "Belos::EpetraMultiVec::MvAddMv call to Update() returned a nonzero value.");
+  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure,
+                     "Belos::EpetraMultiVec::MvAddMv call to Update() returned a nonzero value.");
 }
 
 //-------------------------------------------------------------
@@ -476,21 +476,21 @@ void EpetraMultiVec::MvScale ( const std::vector<double>& alpha )
   // Check to make sure the input vector of scale factors has the same
   // number of entries as the number of columns in the multivector.
   int numvecs = this->NumVectors();
-  TEUCHOS_TEST_FOR_EXCEPTION((int)alpha.size() != numvecs, EpetraMultiVecFailure, 
-		     "Belos::MultiVecTraits<double,Epetra_MultiVec>::MvScale: "
-		     "Vector (alpha) of scaling factors has length " 
-		     << alpha.size() << ", which is different than the number "
-		     "of columns " << numvecs << " in the input multivector "
-		     "(mv).");
+  TEUCHOS_TEST_FOR_EXCEPTION((int)alpha.size() != numvecs, EpetraMultiVecFailure,
+                     "Belos::MultiVecTraits<double,Epetra_MultiVec>::MvScale: "
+                     "Vector (alpha) of scaling factors has length "
+                     << alpha.size() << ", which is different than the number "
+                     "of columns " << numvecs << " in the input multivector "
+                     "(mv).");
   int ret = 0;
   std::vector<int> tmp_index( 1, 0 );
   for (int i=0; i<numvecs; i++) {
     Epetra_MultiVector temp_vec(BELOSEPETRAVIEW, *this, &tmp_index[0], 1);
     ret = temp_vec.Scale( alpha[i] );
-    TEUCHOS_TEST_FOR_EXCEPTION(ret!=0, EpetraMultiVecFailure, 
-		       "Belos::MultiVecTraits<double,Epetra_MultiVec>::MvScale: "
-		       "Call to Epetra_MultiVector::Scale() returned a nonzero "
-		       "error code " << ret << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION(ret!=0, EpetraMultiVecFailure,
+                       "Belos::MultiVecTraits<double,Epetra_MultiVec>::MvScale: "
+                       "Call to Epetra_MultiVector::Scale() returned a nonzero "
+                       "error code " << ret << ".");
     tmp_index[0]++;
   }
 }
@@ -502,38 +502,38 @@ void EpetraMultiVec::MvScale ( const std::vector<double>& alpha )
 //-------------------------------------------------------------
 
 void EpetraMultiVec::MvTransMv ( const double alpha, const MultiVec<double>& A,
-				 Teuchos::SerialDenseMatrix<int,double>& B) const
-{    
+                                 Teuchos::SerialDenseMatrix<int,double>& B) const
+{
   EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A));
-  
+
   if (A_vec) {
     Epetra_LocalMap LocalMap(B.numRows(), 0, Map().Comm());
     Epetra_MultiVector B_Pvec(BELOSEPETRAVIEW, LocalMap, B.values(), B.stride(), B.numCols());
-    
+
     int info = B_Pvec.Multiply( 'T', 'N', alpha, *A_vec, *this, 0.0 );
-    TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure, 
-		       "Belos::EpetraMultiVec::MvTransMv call to Multiply() returned a nonzero value.");
+    TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure,
+                       "Belos::EpetraMultiVec::MvTransMv call to Multiply() returned a nonzero value.");
   }
 }
 
 //-------------------------------------------------------------
 //
 // b[i] = A[i]^T * this[i]
-// 
+//
 //-------------------------------------------------------------
 
 void EpetraMultiVec::MvDot ( const MultiVec<double>& A, std::vector<double>& b ) const
 {
-  EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A)); 
+  EpetraMultiVec *A_vec = dynamic_cast<EpetraMultiVec *>(&const_cast<MultiVec<double> &>(A));
   TEUCHOS_TEST_FOR_EXCEPTION(A_vec==NULL, EpetraMultiVecFailure,
                      "Belos::EpetraMultiVec::MvDot: Dynamic cast from Belos::"
-		     "MultiVec<double> to Belos::EpetraMultiVec failed.");
+                     "MultiVec<double> to Belos::EpetraMultiVec failed.");
   if (A_vec && ( (int)b.size() >= A_vec->NumVectors() ) ) {
      int info = this->Dot( *A_vec, &b[0] );
-     TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure, 
-			"Belos::EpetraMultiVec::MvDot: Call to Epetra_Multi"
-			"Vector::Dot() returned a nonzero error code " 
-			<< info << ".");
+     TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure,
+                        "Belos::EpetraMultiVec::MvDot: Call to Epetra_Multi"
+                        "Vector::Dot() returned a nonzero error code "
+                        << info << ".");
   }
 }
 
@@ -553,20 +553,19 @@ void EpetraMultiVec::MvNorm ( std::vector<double>& normvec, NormType norm_type )
     case ( TwoNorm ) :
       info = Norm2(&normvec[0]);
       break;
-    case ( InfNorm ) :	
+    case ( InfNorm ) :
       info = NormInf(&normvec[0]);
       break;
     default:
-      TEUCHOS_TEST_FOR_EXCEPTION(true, std::invalid_argument, 
-			 "Belos::EpetraMultiVec::MvNorm: Invalid norm_type " 
-			 << norm_type << ".  The current list of valid norm "
-			 "types is {OneNorm, TwoNorm, InfNorm}.");
-      break;
+      TEUCHOS_TEST_FOR_EXCEPTION(true, std::invalid_argument,
+                         "Belos::EpetraMultiVec::MvNorm: Invalid norm_type "
+                         << norm_type << ".  The current list of valid norm "
+                         "types is {OneNorm, TwoNorm, InfNorm}.");
     }
-    TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure, 
-		       "Belos::EpetraMultiVec::MvNorm: Call to Epetra_Multi"
-		       "Vector::Norm() returned a nonzero error code " 
-		       << info << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraMultiVecFailure,
+                       "Belos::EpetraMultiVec::MvNorm: Call to Epetra_Multi"
+                       "Vector::Norm() returned a nonzero error code "
+                       << info << ".");
   }
 }
 
@@ -576,7 +575,7 @@ void EpetraMultiVec::MvNorm ( std::vector<double>& normvec, NormType norm_type )
 //
 ///////////////////////////////////////////////////////////////
 
-EpetraOp::EpetraOp( const Teuchos::RCP<Epetra_Operator> &Op ) 
+EpetraOp::EpetraOp( const Teuchos::RCP<Epetra_Operator> &Op )
   : Epetra_Op(Op)
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -584,15 +583,15 @@ EpetraOp::EpetraOp( const Teuchos::RCP<Epetra_Operator> &Op )
 
   std::ostream& err = *(getErrStream (*Epetra_Op));
   err << "Belos::EpetraOp constructor" << endl;
-  err << "-- On input: Epetra_Op->UseTranspose() = " 
+  err << "-- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 }
 
-void 
-EpetraOp::Apply (const MultiVec<double>& x, 
-		 MultiVec<double>& y, 
-		 ETrans trans) const 
+void
+EpetraOp::Apply (const MultiVec<double>& x,
+                 MultiVec<double>& y,
+                 ETrans trans) const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
   using std::endl;
@@ -601,7 +600,7 @@ EpetraOp::Apply (const MultiVec<double>& x,
   err << "-- Belos::EpetraOp::Apply:" << endl
       << "---- Implements Belos::Operator<double>::Apply()" << endl
       << "---- trans input = " << etransToString (trans) << endl
-      << "---- On input: Epetra_Op->UseTranspose() = " 
+      << "---- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
@@ -609,27 +608,27 @@ EpetraOp::Apply (const MultiVec<double>& x,
   Epetra_MultiVector* vec_x = dynamic_cast<Epetra_MultiVector* >(&temp_x);
   Epetra_MultiVector* vec_y = dynamic_cast<Epetra_MultiVector* >(&y);
 
-  TEUCHOS_TEST_FOR_EXCEPTION(vec_x==NULL || vec_y==NULL, EpetraOpFailure, 
-		     "Belos::EpetraOp::Apply: x and/or y could not be "
-		     "dynamic cast to an Epetra_MultiVector.");
+  TEUCHOS_TEST_FOR_EXCEPTION(vec_x==NULL || vec_y==NULL, EpetraOpFailure,
+                     "Belos::EpetraOp::Apply: x and/or y could not be "
+                     "dynamic cast to an Epetra_MultiVector.");
 
   // Temporarily set the transpose state of Op, if it's not the same
   // as trans, and restore it on exit of this scope.
   EpetraOperatorTransposeScopeGuard guard (*Epetra_Op, trans);
 
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
-  err << "---- Before calling Apply: Epetra_Op->UseTranspose() = " 
+  err << "---- Before calling Apply: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
   // Apply the operator to x and put the result in y.
   const int info = Epetra_Op->Apply (*vec_x, *vec_y);
-  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraOpFailure, 
-		     "Belos::EpetraOp::Apply: The Epetra_Operator's Apply() "
-		     "method returned a nonzero error code of " << info << ".");
+  TEUCHOS_TEST_FOR_EXCEPTION(info!=0, EpetraOpFailure,
+                     "Belos::EpetraOp::Apply: The Epetra_Operator's Apply() "
+                     "method returned a nonzero error code of " << info << ".");
 }
 
-bool 
+bool
 EpetraOp::HasApplyTranspose() const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -647,7 +646,7 @@ EpetraOp::HasApplyTranspose() const
 //
 // ///////////////////////////////////////////////////////////////////
 
-EpetraPrecOp::EpetraPrecOp (const Teuchos::RCP<Epetra_Operator> &Op) 
+EpetraPrecOp::EpetraPrecOp (const Teuchos::RCP<Epetra_Operator> &Op)
   : Epetra_Op(Op)
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -655,17 +654,17 @@ EpetraPrecOp::EpetraPrecOp (const Teuchos::RCP<Epetra_Operator> &Op)
 
   std::ostream& err = *(getErrStream (*Epetra_Op));
   err << "-- Belos::EpetraPrecOp constructor" << endl
-      << "---- On input: Epetra_Op->UseTranspose() = " 
+      << "---- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 }
 
 // The version of Apply() that takes an optional 'trans' argument and
 // returns void implements the Belos::Operator interface.
-void 
-EpetraPrecOp::Apply (const MultiVec<double>& x, 
-		     MultiVec<double>& y, 
-		     ETrans trans) const 
+void
+EpetraPrecOp::Apply (const MultiVec<double>& x,
+                     MultiVec<double>& y,
+                     ETrans trans) const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
   using std::endl;
@@ -674,26 +673,26 @@ EpetraPrecOp::Apply (const MultiVec<double>& x,
   err << "-- Belos::EpetraPrecOp::Apply:" << endl
       << "---- Implements Belos::Operator<double>::Apply()" << endl
       << "---- trans input = " << etransToString (trans) << endl
-      << "---- On input: Epetra_Op->UseTranspose() = " 
+      << "---- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
   MultiVec<double>&  temp_x = const_cast<MultiVec<double> &>(x);
   Epetra_MultiVector* vec_x = dynamic_cast<Epetra_MultiVector* >(&temp_x);
-  TEUCHOS_TEST_FOR_EXCEPTION(vec_x == NULL, EpetraOpFailure, 
-		     "Belos::EpetraPrecOp::Apply: The MultiVec<double> input x "
-		     "cannot be dynamic cast to an Epetra_MultiVector.");
+  TEUCHOS_TEST_FOR_EXCEPTION(vec_x == NULL, EpetraOpFailure,
+                     "Belos::EpetraPrecOp::Apply: The MultiVec<double> input x "
+                     "cannot be dynamic cast to an Epetra_MultiVector.");
   Epetra_MultiVector* vec_y = dynamic_cast<Epetra_MultiVector* >(&y);
-  TEUCHOS_TEST_FOR_EXCEPTION(vec_x == NULL, EpetraOpFailure, 
-		     "Belos::EpetraPrecOp::Apply: The MultiVec<double> input y "
-		     "cannot be dynamic cast to an Epetra_MultiVector.");
+  TEUCHOS_TEST_FOR_EXCEPTION(vec_x == NULL, EpetraOpFailure,
+                     "Belos::EpetraPrecOp::Apply: The MultiVec<double> input y "
+                     "cannot be dynamic cast to an Epetra_MultiVector.");
 
   // Temporarily set the transpose state of Op, if it's not the same
   // as trans, and restore it on exit of this scope.
   EpetraOperatorTransposeScopeGuard guard (*Epetra_Op, trans);
 
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
-  err << "---- Before calling ApplyInverse: Epetra_Op->UseTranspose() = " 
+  err << "---- Before calling ApplyInverse: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
@@ -701,19 +700,19 @@ EpetraPrecOp::Apply (const MultiVec<double>& x,
   // underlying operator.  This may not succeed for all
   // implementations of Epetra_Operator, so we have to check.
   const int info = Epetra_Op->ApplyInverse (*vec_x, *vec_y);
-  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure, 
-		     "Belos::EpetraPrecOp::Apply: Calling ApplyInverse() on the "
-		     "underlying Epetra_Operator object failed, returning a "
-		     "nonzero error code of " << info << ".  This probably means"
-		     " that the underlying Epetra_Operator object doesn't know "
-		     "how to apply its inverse.");
+  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure,
+                     "Belos::EpetraPrecOp::Apply: Calling ApplyInverse() on the "
+                     "underlying Epetra_Operator object failed, returning a "
+                     "nonzero error code of " << info << ".  This probably means"
+                     " that the underlying Epetra_Operator object doesn't know "
+                     "how to apply its inverse.");
 }
 
 // The version of Apply() that takes two arguments and returns int
 // implements the Epetra_Operator interface.
-int 
-EpetraPrecOp::Apply (const Epetra_MultiVector &X, 
-		     Epetra_MultiVector &Y) const 
+int
+EpetraPrecOp::Apply (const Epetra_MultiVector &X,
+                     Epetra_MultiVector &Y) const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
   using std::endl;
@@ -721,7 +720,7 @@ EpetraPrecOp::Apply (const Epetra_MultiVector &X,
   std::ostream& err = *(getErrStream (*Epetra_Op));
   err << "-- Belos::EpetraPrecOp::Apply:" << endl
       << "---- Implements Epetra_Operator::Apply()" << endl
-      << "---- On input: Epetra_Op->UseTranspose() = " 
+      << "---- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl
       << "---- Calling Epetra_Op->ApplyInverse (X, Y)" << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -737,9 +736,9 @@ EpetraPrecOp::Apply (const Epetra_MultiVector &X,
 }
 
 // This implements Epetra_Operator::ApplyInverse().
-int 
-EpetraPrecOp::ApplyInverse (const Epetra_MultiVector &X, 
-			    Epetra_MultiVector &Y) const
+int
+EpetraPrecOp::ApplyInverse (const Epetra_MultiVector &X,
+                            Epetra_MultiVector &Y) const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
   using std::endl;
@@ -747,7 +746,7 @@ EpetraPrecOp::ApplyInverse (const Epetra_MultiVector &X,
   std::ostream& err = *(getErrStream (*Epetra_Op));
   err << "-- Belos::EpetraPrecOp::ApplyInverse:" << endl
       << "---- Implements Epetra_Operator::ApplyInverse()" << endl
-      << "---- On input: Epetra_Op->UseTranspose() = " 
+      << "---- On input: Epetra_Op->UseTranspose() = "
       << Epetra_Op->UseTranspose() << endl
       << "---- Calling Epetra_Op->Apply (X, Y)" << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -762,7 +761,7 @@ EpetraPrecOp::ApplyInverse (const Epetra_MultiVector &X,
   return info;
 }
 
-bool 
+bool
 EpetraPrecOp::HasApplyTranspose() const
 {
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
@@ -775,26 +774,26 @@ EpetraPrecOp::HasApplyTranspose() const
 
 
 // ///////////////////////////////////////////////////////////////////
-// 
+//
 // Specialization of Belos::OperatorTraits for Epetra_Operator.
 //
 // ///////////////////////////////////////////////////////////////////
-  
+
 void
 OperatorTraits<double, Epetra_MultiVector, Epetra_Operator>::
-Apply (const Epetra_Operator& Op, 
-       const Epetra_MultiVector& x, 
+Apply (const Epetra_Operator& Op,
+       const Epetra_MultiVector& x,
        Epetra_MultiVector& y,
        ETrans trans)
-{ 
+{
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
   using std::endl;
 
   std::ostream& err = *(getErrStream (Op));
-  err << "Belos::OperatorTraits<double, Epetra_MultiVector, Epetra_Operator>::Apply:" 
+  err << "Belos::OperatorTraits<double, Epetra_MultiVector, Epetra_Operator>::Apply:"
       << endl
       << "-- trans input = " << etransToString (trans) << endl
-      << "-- On input: Op.UseTranspose() = " 
+      << "-- On input: Op.UseTranspose() = "
       << Op.UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
@@ -803,16 +802,16 @@ Apply (const Epetra_Operator& Op,
   EpetraOperatorTransposeScopeGuard guard (Op, trans);
 
 #ifdef BELOS_EPETRA_AGGRESSIVE_DEBUGGING
-  err << "-- Before calling Op.Apply: Op.UseTranspose() = " 
+  err << "-- Before calling Op.Apply: Op.UseTranspose() = "
       << Op.UseTranspose() << endl;
 #endif // BELOS_EPETRA_AGGRESSIVE_DEBUGGING
 
   const int info = Op.Apply (x, y);
-  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure, 
-		     "Belos::OperatorTraits::Apply (Epetra specialization): "
-		     "Calling the Epetra_Operator object's Apply() method "
-		     "failed, returning a nonzero error code of " << info
-		     << ".");
+  TEUCHOS_TEST_FOR_EXCEPTION(info != 0, EpetraOpFailure,
+                     "Belos::OperatorTraits::Apply (Epetra specialization): "
+                     "Calling the Epetra_Operator object's Apply() method "
+                     "failed, returning a nonzero error code of " << info
+                     << ".");
 }
 
 bool

--- a/packages/belos/tpetra/test/Issue_3235.cpp
+++ b/packages/belos/tpetra/test/Issue_3235.cpp
@@ -36,9 +36,11 @@ int main(int argc, char *argv[]) {
       uniqueMapArray[i] = i;
     }
 
-    RCP<const map_type> UniqueMap = rcp (new map_type (-1, uniqueMapArray, 0, TeuchosComm));
+    const Tpetra::global_size_t INVALID =
+      Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid();
+    RCP<const map_type> UniqueMap = rcp (new map_type (INVALID, uniqueMapArray, 0, TeuchosComm));
 
-    RCP<crs_matrix_type> K = rcp (new crs_matrix_type (UniqueMap,10));
+    RCP<crs_matrix_type> K = rcp (new crs_matrix_type (UniqueMap, 10));
 
     for (LO i = 0; i < static_cast<LO>(UniqueMap->getNodeNumElements()); ++i) {
       LO numEntries = 10-i;

--- a/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp
+++ b/packages/teuchos/numerics/src/Teuchos_SerialDenseMatrix.hpp
@@ -34,8 +34,6 @@
 // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
-// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
-//
 // ***********************************************************************
 // @HEADER
 
@@ -52,7 +50,7 @@
 #include "Teuchos_ConfigDefs.hpp"
 #include "Teuchos_Assert.hpp"
 #include "Teuchos_SerialSymDenseMatrix.hpp"
-
+#include <cstddef>
 #include <utility>
 
 /*!     \class Teuchos::SerialDenseMatrix
@@ -82,7 +80,7 @@ public:
   /*! Creates a empty matrix of no dimension.  The Shaping methods should be used to size this matrix.
     Values of this matrix should be set using the [], (), or = operators.
   */
-  SerialDenseMatrix();
+  SerialDenseMatrix() = default;
 
   //! Shaped Constructor
   /*!
@@ -244,8 +242,6 @@ public:
     A[j][i] will return the same element as A(i,j).
 
     \return Pointer to the ScalarType array at the \c colIndex column ( \c values_+colIndex*stride_ ).
-    \warning The validity of \c colIndex will only be checked if Teuchos is     configured with
-    --enable-teuchos-abc.
   */
   ScalarType* operator [] (OrdinalType colIndex);
 
@@ -254,14 +250,12 @@ public:
     A[j][i] will return the same element as A(i,j).
 
     \return Pointer to the ScalarType array at the \c colIndex column ( \c values_+colIndex*stride_ ).
-    \warning The validity of \c colIndex will only be checked if Teuchos is     configured with
-    --enable-teuchos-abc.
   */
   const ScalarType* operator [] (OrdinalType colIndex) const;
 
   //! Data array access method.
   /*! \return Pointer to the ScalarType data array contained in the object. */
-  ScalarType* values() const { return(values_); }
+  ScalarType* values() const { return values_; }
 
   //@}
 
@@ -398,12 +392,23 @@ protected:
     ScalarType alpha = ScalarTraits<ScalarType>::zero() );
   void deleteArrays();
   void checkIndex( OrdinalType rowIndex, OrdinalType colIndex = 0 ) const;
-  OrdinalType numRows_;
-  OrdinalType numCols_;
-  OrdinalType stride_;
-  bool valuesCopied_;
-  ScalarType* values_;
 
+  static ScalarType*
+  allocateValues(const OrdinalType numRows,
+                 const OrdinalType numCols)
+  {
+    const size_t size = size_t(numRows) * size_t(numCols);
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wvla"
+    return new ScalarType[size];
+#pragma GCC diagnostic pop
+  }
+
+  OrdinalType numRows_ = 0;
+  OrdinalType numCols_ = 0;
+  OrdinalType stride_ = 0;
+  bool valuesCopied_ = false;
+  ScalarType* values_ = nullptr;
 }; // class Teuchos_SerialDenseMatrix
 
 //----------------------------------------------------------------------------------------------------
@@ -411,20 +416,18 @@ protected:
 //----------------------------------------------------------------------------------------------------
 
 template<typename OrdinalType, typename ScalarType>
-SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix()
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(0), numCols_(0), stride_(0), valuesCopied_(false), values_(0)
-{}
-
-template<typename OrdinalType, typename ScalarType>
 SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   OrdinalType numRows_in, OrdinalType numCols_in, bool zeroOut
   )
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(numRows_in), numCols_(numCols_in), stride_(numRows_in)
+  : numRows_(numRows_in),
+    numCols_(numCols_in),
+    stride_(numRows_in),
+    valuesCopied_(true),
+    values_(allocateValues(numRows_in, numCols_in))
 {
-  values_ = new ScalarType[stride_*numCols_];
-  valuesCopied_ = true;
-  if (zeroOut == true)
+  if (zeroOut) {
     putScalar();
+  }
 }
 
 template<typename OrdinalType, typename ScalarType>
@@ -432,21 +435,24 @@ SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   DataAccess CV, ScalarType* values_in, OrdinalType stride_in, OrdinalType numRows_in,
   OrdinalType numCols_in
   )
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), numRows_(numRows_in), numCols_(numCols_in), stride_(stride_in),
-    valuesCopied_(false), values_(values_in)
+  : numRows_(numRows_in),
+    numCols_(numCols_in),
+    stride_(stride_in),
+    valuesCopied_(false),
+    values_(values_in)
 {
   if(CV == Copy)
   {
     stride_ = numRows_;
-    values_ = new ScalarType[stride_*numCols_];
+    values_ = allocateValues(stride_, numCols_);
     copyMat(values_in, stride_in, numRows_, numCols_, values_, stride_, 0, 0);
     valuesCopied_ = true;
   }
 }
 
 template<typename OrdinalType, typename ScalarType>
-SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseMatrix<OrdinalType, ScalarType> &Source, ETransp trans) 
-  : CompObject(),BLAS<OrdinalType,ScalarType>(), numRows_(0), numCols_(0), stride_(0), valuesCopied_(true), values_(0)
+SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseMatrix<OrdinalType, ScalarType> &Source, ETransp trans)
+  : valuesCopied_(true)
 {
   if ( trans == Teuchos::NO_TRANS )
   {
@@ -462,9 +468,8 @@ SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseM
     else
     {
       stride_ = numRows_;
-      const OrdinalType newsize = stride_ * numCols_;
-      if(newsize > 0) {
-        values_ = new ScalarType[newsize];
+      if(stride_ > 0 && numCols_ > 0) {
+        values_ = allocateValues(stride_, numCols_);
         copyMat(Source.values_, Source.stride_, numRows_, numCols_, values_, stride_, 0, 0);
       }
       else {
@@ -478,7 +483,7 @@ SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseM
     numRows_ = Source.numCols_;
     numCols_ = Source.numRows_;
     stride_ = numRows_;
-    values_ = new ScalarType[stride_*numCols_];
+    values_ = allocateValues(stride_, numCols_);
     for (OrdinalType j=0; j<numCols_; j++) {
       for (OrdinalType i=0; i<numRows_; i++) {
         values_[j*stride_ + i] = Teuchos::ScalarTraits<ScalarType>::conjugate(Source.values_[i*Source.stride_ + j]);
@@ -490,7 +495,7 @@ SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(const SerialDenseM
     numRows_ = Source.numCols_;
     numCols_ = Source.numRows_;
     stride_ = numRows_;
-    values_ = new ScalarType[stride_*numCols_];
+    values_ = allocateValues(stride_, numCols_);
     for (OrdinalType j=0; j<numCols_; j++) {
       for (OrdinalType i=0; i<numRows_; i++) {
         values_[j*stride_ + i] = Source.values_[i*Source.stride_ + j];
@@ -504,13 +509,13 @@ template<typename OrdinalType, typename ScalarType>
 SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   DataAccess CV, const SerialDenseMatrix<OrdinalType, ScalarType> &Source
   )
-  : CompObject(), numRows_(Source.numRows_), numCols_(Source.numCols_), stride_(Source.stride_),
+  : numRows_(Source.numRows_), numCols_(Source.numCols_), stride_(Source.stride_),
     valuesCopied_(false), values_(Source.values_)
 {
   if(CV == Copy)
   {
     stride_ = numRows_;
-    values_ = new ScalarType[stride_ * numCols_];
+    values_ = allocateValues(stride_, numCols_);
     copyMat(Source.values_, Source.stride_, numRows_, numCols_, values_, stride_, 0, 0);
     valuesCopied_ = true;
   }
@@ -529,7 +534,7 @@ SerialDenseMatrix<OrdinalType, ScalarType>::SerialDenseMatrix(
   if(CV == Copy)
   {
     stride_ = numRows_in;
-    values_ = new ScalarType[stride_ * numCols_in];
+    values_ = allocateValues(stride_, numCols_in);
     copyMat(Source.values_, Source.stride_, numRows_in, numCols_in, values_, stride_, startRow, startCol);
     valuesCopied_ = true;
   }
@@ -558,7 +563,7 @@ int SerialDenseMatrix<OrdinalType, ScalarType>::shape(
   numRows_ = numRows_in;
   numCols_ = numCols_in;
   stride_ = numRows_;
-  values_ = new ScalarType[stride_*numCols_];
+  values_ = allocateValues(stride_, numCols_);
   putScalar();
   valuesCopied_ = true;
   return(0);
@@ -573,7 +578,7 @@ int SerialDenseMatrix<OrdinalType, ScalarType>::shapeUninitialized(
   numRows_ = numRows_in;
   numCols_ = numCols_in;
   stride_ = numRows_;
-  values_ = new ScalarType[stride_*numCols_];
+  values_ = allocateValues(stride_, numCols_);
   valuesCopied_ = true;
   return(0);
 }
@@ -584,7 +589,7 @@ int SerialDenseMatrix<OrdinalType, ScalarType>::reshape(
   )
 {
   // Allocate space for new matrix
-  ScalarType* values_tmp = new ScalarType[numRows_in * numCols_in];
+  ScalarType* values_tmp = allocateValues(numRows_in, numCols_in);
   ScalarType zero = ScalarTraits<ScalarType>::zero();
   for(OrdinalType k = 0; k < numRows_in * numCols_in; k++)
   {
@@ -691,9 +696,8 @@ SerialDenseMatrix<OrdinalType, ScalarType>::operator=(
       numRows_ = Source.numRows_;
       numCols_ = Source.numCols_;
       stride_ = Source.numRows_;
-      const OrdinalType newsize = stride_ * numCols_;
-      if(newsize > 0) {
-        values_ = new ScalarType[newsize];
+      if(stride_ > 0 && numCols_ > 0) {
+        values_ = allocateValues(stride_, numCols_);
         valuesCopied_ = true;
       }
       else {
@@ -711,9 +715,8 @@ SerialDenseMatrix<OrdinalType, ScalarType>::operator=(
         numRows_ = Source.numRows_;
         numCols_ = Source.numCols_;
         stride_ = Source.numRows_;
-        const OrdinalType newsize = stride_ * numCols_;
-        if(newsize > 0) {
-          values_ = new ScalarType[newsize];
+        if(stride_ > 0 && numCols_ > 0) {
+          values_ = allocateValues(stride_, numCols_);
           valuesCopied_ = true;
         }
       }
@@ -1096,7 +1099,7 @@ std::ostream& operator<< (std::ostream& os, const Teuchos::SerialDenseMatrix<Ord
 }
 #endif
 
-/// \brief Ostream manipulator for SerialDenseMatrix 
+/// \brief Ostream manipulator for SerialDenseMatrix
 template<typename OrdinalType, typename ScalarType>
 struct SerialDenseMatrixPrinter {
 public:
@@ -1106,7 +1109,7 @@ public:
       : obj(obj_in) {}
 };
 
-/// \brief Output SerialDenseMatrix object through its stream manipulator. 
+/// \brief Output SerialDenseMatrix object through its stream manipulator.
 template<typename OrdinalType, typename ScalarType>
 std::ostream&
 operator<<(std::ostream &out,

--- a/packages/teuchos/numerics/src/Teuchos_SerialSymDenseMatrix.hpp
+++ b/packages/teuchos/numerics/src/Teuchos_SerialSymDenseMatrix.hpp
@@ -35,8 +35,6 @@
 // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
-// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
-//
 // ***********************************************************************
 // @HEADER
 
@@ -140,7 +138,7 @@ class SerialSymDenseMatrix : public CompObject, public BLAS<OrdinalType,ScalarTy
     Note: By default the active part of the matrix is assumed to be the lower triangular part.
     To set the upper part as active, call SetUpper(). See Detailed Description section for further discussion.
    */
-  SerialSymDenseMatrix();
+  SerialSymDenseMatrix() = default;
 
   //! Basic constructor; defines a matrix of \c numRowsCols size and (optionally) initializes it.
   /*!
@@ -427,14 +425,24 @@ class SerialSymDenseMatrix : public CompObject, public BLAS<OrdinalType,ScalarTy
 
   void deleteArrays();
   void checkIndex( OrdinalType rowIndex, OrdinalType colIndex = 0 ) const;
-  OrdinalType numRowCols_;
-  OrdinalType stride_;
-  bool valuesCopied_;
-  ScalarType* values_;
-  bool upper_;
-  char UPLO_;
 
+  static ScalarType*
+  allocateValues(const OrdinalType numRows,
+                 const OrdinalType numCols)
+  {
+    const size_t size = size_t(numRows) * size_t(numCols);
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wvla"
+    return new ScalarType[size];
+#pragma GCC diagnostic pop
+  }
 
+  OrdinalType numRowCols_ = 0;
+  OrdinalType stride_ = 0;
+  bool valuesCopied_ = false;
+  ScalarType* values_ = nullptr;
+  bool upper_ = false;
+  char UPLO_ {'L'};
 };
 
 //----------------------------------------------------------------------------------------------------
@@ -442,28 +450,22 @@ class SerialSymDenseMatrix : public CompObject, public BLAS<OrdinalType,ScalarTy
 //----------------------------------------------------------------------------------------------------
 
 template<typename OrdinalType, typename ScalarType>
-SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix()
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), 
-    numRowCols_(0), stride_(0), valuesCopied_(false), values_(0), upper_(false), UPLO_('L')
-{}
-
-template<typename OrdinalType, typename ScalarType>
 SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(OrdinalType numRowCols_in, bool zeroOut)
-  : CompObject(), BLAS<OrdinalType,ScalarType>(),
-    numRowCols_(numRowCols_in), stride_(numRowCols_in), valuesCopied_(false), upper_(false), UPLO_('L')
+  : numRowCols_(numRowCols_in), stride_(numRowCols_in), valuesCopied_(false), upper_(false), UPLO_('L')
 {
-  values_ = new ScalarType[stride_*numRowCols_];
+  values_ = allocateValues(stride_, numRowCols_);
   valuesCopied_ = true;
-  if (zeroOut == true)
-    putScalar( Teuchos::ScalarTraits<ScalarType>::zero(), true );
+  if (zeroOut) {
+    const ScalarType ZERO = Teuchos::ScalarTraits<ScalarType>::zero();
+    putScalar(ZERO, true);
+  }
 }
 
 template<typename OrdinalType, typename ScalarType>
 SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(
   DataAccess CV, bool upper_in, ScalarType* values_in, OrdinalType stride_in, OrdinalType numRowCols_in
   )
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), 
-    numRowCols_(numRowCols_in), stride_(stride_in), valuesCopied_(false),
+  : numRowCols_(numRowCols_in), stride_(stride_in), valuesCopied_(false),
     values_(values_in), upper_(upper_in)
 {
   if (upper_)
@@ -474,16 +476,15 @@ SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(
   if(CV == Copy)
   {
     stride_ = numRowCols_;
-    values_ = new ScalarType[stride_*numRowCols_];
+    values_ = allocateValues(stride_, numRowCols_);
     copyMat(upper_in, values_in, stride_in, numRowCols_, upper_, values_, stride_, 0);
     valuesCopied_ = true;
   }
 }
 
 template<typename OrdinalType, typename ScalarType>
-SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(const SerialSymDenseMatrix<OrdinalType, ScalarType> &Source) 
-  : CompObject(), BLAS<OrdinalType,ScalarType>(), 
-    numRowCols_(Source.numRowCols_), stride_(0), valuesCopied_(true), 
+SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(const SerialSymDenseMatrix<OrdinalType, ScalarType> &Source)
+  : numRowCols_(Source.numRowCols_), stride_(0), valuesCopied_(true),
     values_(0), upper_(Source.upper_), UPLO_(Source.UPLO_)
 {
   if (!Source.valuesCopied_)
@@ -495,13 +496,13 @@ SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(const Serial
   else
   {
     stride_ = numRowCols_;
-    const OrdinalType newsize = stride_ * numRowCols_;
-    if(newsize > 0) {
-      values_ = new ScalarType[newsize];
+    if(stride_ > 0 && numRowCols_ > 0) {
+      values_ = allocateValues(stride_, numRowCols_);
       copyMat(Source.upper_, Source.values_, Source.stride_, numRowCols_, upper_, values_, stride_, 0);
     }
     else {
-      numRowCols_ = 0; stride_ = 0;
+      numRowCols_ = 0;
+      stride_ = 0;
       valuesCopied_ = false;
     }
   }
@@ -517,7 +518,7 @@ SerialSymDenseMatrix<OrdinalType, ScalarType>::SerialSymDenseMatrix(
   if(CV == Copy)
   {
     stride_ = numRowCols_in;
-    values_ = new ScalarType[stride_ * numRowCols_in];
+    values_ = allocateValues(stride_, numRowCols_in);
     copyMat(Source.upper_, Source.values_, Source.stride_, numRowCols_in, upper_, values_, stride_, startRowCol);
     valuesCopied_ = true;
   }
@@ -543,7 +544,7 @@ int SerialSymDenseMatrix<OrdinalType, ScalarType>::shape( OrdinalType numRowCols
   deleteArrays(); // Get rid of anything that might be already allocated
   numRowCols_ = numRowCols_in;
   stride_ = numRowCols_;
-  values_ = new ScalarType[stride_*numRowCols_];
+  values_ = allocateValues(stride_, numRowCols_);
   putScalar( Teuchos::ScalarTraits<ScalarType>::zero(), true );
   valuesCopied_ = true;
   return(0);
@@ -555,7 +556,7 @@ int SerialSymDenseMatrix<OrdinalType, ScalarType>::shapeUninitialized( OrdinalTy
   deleteArrays(); // Get rid of anything that might be already allocated
   numRowCols_ = numRowCols_in;
   stride_ = numRowCols_;
-  values_ = new ScalarType[stride_*numRowCols_];
+  values_ = allocateValues(stride_, numRowCols_);
   valuesCopied_ = true;
   return(0);
 }
@@ -564,8 +565,8 @@ template<typename OrdinalType, typename ScalarType>
 int SerialSymDenseMatrix<OrdinalType, ScalarType>::reshape( OrdinalType numRowCols_in )
 {
   // Allocate space for new matrix
-  ScalarType* values_tmp = new ScalarType[numRowCols_in * numRowCols_in];
-  ScalarType zero = ScalarTraits<ScalarType>::zero();
+  ScalarType* values_tmp = allocateValues(numRowCols_in, numRowCols_in);
+  const ScalarType zero = Teuchos::ScalarTraits<ScalarType>::zero();
   for(OrdinalType k = 0; k < numRowCols_in * numRowCols_in; k++)
   {
     values_tmp[k] = zero;
@@ -717,9 +718,8 @@ SerialSymDenseMatrix<OrdinalType, ScalarType>::operator=( const SerialSymDenseMa
       stride_ = Source.numRowCols_;
       upper_ = Source.upper_;
       UPLO_ = Source.UPLO_;
-      const OrdinalType newsize = stride_ * numRowCols_;
-      if(newsize > 0) {
-        values_ = new ScalarType[newsize];
+      if(stride_ > 0 && numRowCols_ > 0) {
+        values_ = allocateValues(stride_, numRowCols_);
         valuesCopied_ = true;
       }
       else {
@@ -739,9 +739,8 @@ SerialSymDenseMatrix<OrdinalType, ScalarType>::operator=( const SerialSymDenseMa
         stride_ = Source.numRowCols_;
         upper_ = Source.upper_;
         UPLO_ = Source.UPLO_;
-        const OrdinalType newsize = stride_ * numRowCols_;
-        if(newsize > 0) {
-          values_ = new ScalarType[newsize];
+        if(stride_ > 0 && numRowCols_ > 0) {
+          values_ = allocateValues(stride_, numRowCols_);
           valuesCopied_ = true;
         }
       }
@@ -1196,7 +1195,7 @@ std::ostream& operator<< (std::ostream& os, const Teuchos::SerialSymDenseMatrix<
 }
 #endif
 
-/// \brief Ostream manipulator for SerialSymDenseMatrix 
+/// \brief Ostream manipulator for SerialSymDenseMatrix
 template<typename OrdinalType, typename ScalarType>
 struct SerialSymDenseMatrixPrinter {
 public:
@@ -1206,7 +1205,7 @@ public:
       : obj(obj_in) {}
 };
 
-/// \brief Output SerialSymDenseMatrix object through its stream manipulator. 
+/// \brief Output SerialSymDenseMatrix object through its stream manipulator.
 template<typename OrdinalType, typename ScalarType>
 std::ostream&
 operator<<(std::ostream &out,

